### PR TITLE
Prove txs that span more than two shares

### LINKED ISSUE #621: Add support for returning proofs for transactions larger than two shares
Almost all standard transactions will be able to fit in two shares, currently 512 bytes, but it is possible to create a transaction that is larger than 512 bytes. If this occurs, then the current way we generate inclusion proofs (#615) for those txs will not handle them. This can be fixed by removing some arbitrary restrictions, along with changing the nmt API so that we can create inclusion proofs over a given range.

### PR DESCRIPTION
**Pull request by evan-forbes**
This PR uses the updated nmt to create inclusion proofs for txs that span more than two shares.

This PR is ready for review, but we are waiting for the required changes to be included in nmt.

Closes #621

blocked by https://github.com/celestiaorg/nmt/pull/58

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -8,7 +8,7 @@ require (
 	github.com/adlio/schema v1.3.0
 	github.com/btcsuite/btcd v0.22.0-beta
 	github.com/btcsuite/btcutil v1.0.3-0.20201208143702-a53e38424cce
-	github.com/celestiaorg/nmt v0.8.0
+	github.com/celestiaorg/nmt v0.9.1-0.20220529205217-a1a0a2f56fce
```
Comment (liamsi) [inline go.mod:11]:
```suggestion
	github.com/celestiaorg/nmt v0.9.0
```

File: pkg/prove/proof_test.go [hunk 2]
```diff
@@ -19,7 +19,7 @@ import (
 func TestTxInclusion(t *testing.T) {
 	txCount := 100
 	typicalBlockData := types.Data{
-		Txs:      generateRandomlySizedContiguousShares(txCount, 200),
+		Txs:      generateRandomlySizedContiguousShares(txCount, 1000),
```
Comment (liamsi) [inline pkg/prove/proof_test.go:22]: Can this go up to 4069 or more? Or can we somehow explicitly test the case where a tx/msg spans multiple rows as well?

File: pkg/prove/proof.go [hunk 3]
```diff
@@ -87,7 +91,7 @@ func TxInclusion(codec rsmt2d.Codec, data types.Data, origSquareSize, txIndex ui
 }
 
 // txSharePosition returns the share that a given transaction is included in.
-// returns -1 if index is greater than that of the provided txs.
+// returns and error if index is greater than that of the provided txs.
```
Comment (liamsi) [inline pkg/prove/proof.go:94]:
```suggestion
// returns an error if index is greater than that of the provided txs.
```

File: pkg/prove/proof.go [hunk 4]
```diff
@@ -53,16 +52,21 @@ func TxInclusion(codec rsmt2d.Codec, data types.Data, origSquareSize, txIndex ui
...
+		}
+		// if this is not the last row, then select for the rest of the row
+		if i != (len(rowShares) - 1) {
+			endLeafPos = origSquareSize - 1
 		}
 
-		shares = append(shares, row[pos])
+		shares = append(shares, row[startLeafPos:endLeafPos+1]...)
 
-		proof, err := tree.Prove(int(pos))
```
Comment (liamsi) [inline pkg/prove/proof.go:65]: TBH, even without the range api, couldn't we just have added all share proofs individually as before?

**Review by liamsi — COMMENTED** [review]
Let's bring this over the finish line

File: go.mod (hunk 1 above)
Comment (evan-forbes) [reply go.mod:11]: We do still have to change this, but to v0.10.0 after we merge https://github.com/celestiaorg/nmt/pull/58

I was incorrect when I said that v0.9.0 would be sufficient, we also have to be able to verify proofs that span more than one nmt leaf.

File: pkg/prove/proof.go (hunk 4 above)
Comment (evan-forbes) [reply pkg/prove/proof.go:65]: yeah, we could have done that as well. It wouldn't have been that much worse of a solution either for 99% of cases. buuuuut now we have a good solution and extended nmt's functionality :slightly_smiling_face:

File: go.mod (hunk 1 above)
Comment (liamsi) [reply go.mod:11]: Oh, I missed that one. Will have a 👀

Comment (liamsi) [reply go.mod:11]: approooooved!

File: pkg/prove/proof_test.go (hunk 2 above)
Comment (evan-forbes) [reply pkg/prove/proof_test.go]: yep, good idea. https://github.com/celestiaorg/celestia-core/blob/ff2bcbde0685c8a311d087514bb009e3cc9e6bac/pkg/prove/proof_test.go#L29-L47

**Comment by evan-forbes** [comment]
forgot to mark ready for review... we just have to rerun the CI until it passes, then we should be good to go

File: pkg/prove/proof.go [hunk 5]
```diff
@@ -42,8 +41,10 @@ func TxInclusion(codec rsmt2d.Codec, data types.Data, origSquareSize, txIndex ui
 	var rowRoots []tmbytes.HexBytes //nolint:prealloc // rarely will this contain more than a single root
 	for i, row := range rowShares {
 		// create an nmt to use to generate a proof
-		tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(origSquareSize))
+		fmt.Println("square size", data.OriginalSquareSize, len(row), "rows", startRow, endRow, i, "leaves", startLeaf, endLeaf, "positions", startPos, endPos, "txindex", txIndex)
```
Comment (liamsi) [inline pkg/prove/proof.go:44]: Println?

File: pkg/prove/proof.go [hunk 6]
```diff
@@ -42,8 +41,10 @@ func TxInclusion(codec rsmt2d.Codec, data types.Data, origSquareSize, txIndex ui
...
+		tree := wrapper.NewErasuredNamespacedMerkleTree(data.OriginalSquareSize)
 		for j, share := range row {
+			fmt.Println(j, share[:25])
```
Comment (liamsi) [inline pkg/prove/proof.go:47]: Looks like you forgot some debug output.

Comment (evan-forbes) [reply pkg/prove/proof.go]: :man_facepalming:

**Review by liamsi — APPROVED** [review]
Nice bug squashing 🐛 🛑

**Pull request merged**